File: packages/semantic-cache-py/betterdb_semantic_cache/telemetry.py
from __future__ import annotations

import weakref
from dataclasses import dataclass
from typing import Any

from opentelemetry import trace
from opentelemetry.trace import Tracer
from prometheus_client import REGISTRY as _DEFAULT_REGISTRY
from prometheus_client import CollectorRegistry, Counter, Histogram

_OPERATION_BUCKETS = [0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0]
_SIMILARITY_BUCKETS = [0.02, 0.05, 0.08, 0.1, 0.12, 0.15, 0.2, 0.3, 0.5, 1.0, 2.0]

# WeakKeyDictionary keyed on the registry object itself so entries are
# automatically evicted when the registry is garbage-collected, preventing
# stale metric instances from being returned to a new registry that happens
# to occupy the same memory address.
_metric_cache: weakref.WeakKeyDictionary[CollectorRegistry, dict[str, Any]] = (
    weakref.WeakKeyDictionary()
)


def _get_or_create_counter(
    registry: CollectorRegistry,
    name: str,
    documentation: str,
    labelnames: list[str],
) -> Counter:
    by_name = _metric_cache.setdefault(registry, {})
    if name not in by_name:
        try:
            by_name[name] = Counter(name, documentation, labelnames, registry=registry)
        except ValueError:
            # Metric already registered externally — retrieve via the private
            # _names_to_collectors map (guarded against removal in future releases).
            existing = getattr(registry, "_names_to_collectors", {}).get(name)
            if existing is None:
                raise
            by_name[name] = existing
    return by_name[name]  # type: ignore[return-value]


def _get_or_create_histogram(
    registry: CollectorRegistry,
    name: str,
    documentation: str,
    labelnames: list[str],
    buckets: list[float],
) -> Histogram:
    by_name = _metric_cache.setdefault(registry, {})
    if name not in by_name:
        try:
            by_name[name] = Histogram(
                name, documentation, labelnames, buckets=buckets, registry=registry
            )
        except ValueError:
            # Metric already registered externally — retrieve via the private
            # _names_to_collectors map (guarded against removal in future releases).
            existing = getattr(registry, "_names_to_collectors", {}).get(name)
            if existing is None:
                raise
            by_name[name] = existing
    return by_name[name]  # type: ignore[return-value]


@dataclass
class SemanticCacheMetrics:
    requests_total: Counter
    similarity_score: Histogram
    operation_duration: Histogram
    embedding_duration: Histogram
    cost_saved_total: Counter
    embedding_cache_total: Counter
    stale_model_evictions: Counter
    config_refresh_failed: Counter
    discovery_write_failed: Counter
    judge_decisions_total: Counter
    judge_duration_seconds: Histogram


@dataclass
class Telemetry:
    tracer: Tracer
    metrics: SemanticCacheMetrics


def create_telemetry(
    prefix: str,
    tracer_name: str,
    registry: CollectorRegistry | None = None,
) -> Telemetry:
    reg = registry or _DEFAULT_REGISTRY
    tracer = trace.get_tracer(tracer_name)

    metrics = SemanticCacheMetrics(
        requests_total=_get_or_create_counter(
            reg,
            f"{prefix}_requests_total",
            "Total number of semantic cache requests",
            ["cache_name", "result", "category"],
        ),
        similarity_score=_get_or_create_histogram(
            reg,
            f"{prefix}_similarity_score",
            "Cosine distance similarity scores for cache lookups",
            ["cache_name", "category"],
            _SIMILARITY_BUCKETS,
        ),
        operation_duration=_get_or_create_histogram(
            reg,
            f"{prefix}_operation_duration_seconds",
            "Duration of semantic cache operations in seconds",
            ["cache_name", "operation"],
            _OPERATION_BUCKETS,
        ),
        embedding_duration=_get_or_create_histogram(
            reg,
            f"{prefix}_embedding_duration_seconds",
            "Duration of embedding function calls in seconds",
            ["cache_name"],
            _OPERATION_BUCKETS,
        ),
        cost_saved_total=_get_or_create_counter(
            reg,
            f"{prefix}_cost_saved_total",
            "Estimated cost saved in dollars from semantic cache hits",
            ["cache_name", "category"],
        ),
        embedding_cache_total=_get_or_create_counter(
            reg,
            f"{prefix}_embedding_cache_total",
            "Total embedding cache lookups (hit or miss)",
            ["cache_name", "result"],
        ),
        stale_model_evictions=_get_or_create_counter(
            reg,
            f"{prefix}_stale_model_evictions_total",
            "Entries evicted due to stale_after_model_change detection",
            ["cache_name"],
        ),
        config_refresh_failed=_get_or_create_counter(
            reg,
            f"{prefix}_config_refresh_failed_total",
            "Count of failed periodic config refreshes (HGETALL on __config).",
            ["cache_name"],
        ),
        discovery_write_failed=_get_or_create_counter(
            reg,
            f"{prefix}_discovery_write_failed_total",
            "Count of failed discovery marker writes (HSET registry, SET heartbeat).",
            ["cache_name"],
        ),
        judge_decisions_total=_get_or_create_counter(
            reg,
            f"{prefix}_judge_decisions_total",
            "LLM-as-judge decisions by outcome.",
            ["cache_name", "category", "decision"],
        ),
        judge_duration_seconds=_get_or_create_histogram(
            reg,
            f"{prefix}_judge_duration_seconds",
            "Duration of LLM-as-judge calls in seconds.",
            ["cache_name", "category", "decision"],
            [0.05, 0.1, 0.25, 0.5, 1, 2, 5],
        ),
    )

    return Telemetry(tracer=tracer, metrics=metrics)

File: packages/semantic-cache/src/telemetry.ts
import { trace, type Tracer } from '@opentelemetry/api';
import {
  Counter,
  Histogram,
  Registry,
  register as defaultRegistry,
  type CounterConfiguration,
  type HistogramConfiguration,
} from 'prom-client';

interface TelemetryFactoryOptions {
  prefix: string;
  tracerName: string;
  registry?: Registry;
}

interface CacheMetrics {
  requestsTotal: Counter;
  similarityScore: Histogram;
  operationDuration: Histogram;
  embeddingDuration: Histogram;
  costSavedTotal: Counter;
  embeddingCacheTotal: Counter;
  staleModelEvictions: Counter;
  discoveryWriteFailed: Counter;
  configRefreshFailed: Counter;
  judgeDecisions: Counter;
  judgeDuration: Histogram;
}

export interface Telemetry {
  tracer: Tracer;
  metrics: CacheMetrics;
}

function getOrCreateCounter(
  registry: Registry,
  config: CounterConfiguration<string>,
): Counter {
  const existing = registry.getSingleMetric(config.name);
  if (existing) return existing as Counter;
  return new Counter({ ...config, registers: [registry] });
}

function getOrCreateHistogram(
  registry: Registry,
  config: HistogramConfiguration<string>,
): Histogram {
  const existing = registry.getSingleMetric(config.name);
  if (existing) return existing as Histogram;
  return new Histogram({ ...config, registers: [registry] });
}

export function createTelemetry(opts: TelemetryFactoryOptions): Telemetry {
  const registry = opts.registry ?? defaultRegistry;
  const tracer = trace.getTracer(opts.tracerName);

  const operationBuckets = [0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0];

  const requestsTotal = getOrCreateCounter(registry, {
    name: `${opts.prefix}_requests_total`,
    help: 'Total number of semantic cache requests',
    labelNames: ['cache_name', 'result', 'category'],
  });

  const similarityScore = getOrCreateHistogram(registry, {
    name: `${opts.prefix}_similarity_score`,
    help: 'Cosine distance similarity scores for cache lookups',
    labelNames: ['cache_name', 'category'],
    buckets: [0.02, 0.05, 0.08, 0.1, 0.12, 0.15, 0.2, 0.3, 0.5, 1.0, 2.0],
  });

  const operationDuration = getOrCreateHistogram(registry, {
    name: `${opts.prefix}_operation_duration_seconds`,
    help: 'Duration of semantic cache operations in seconds',
    labelNames: ['cache_name', 'operation'],
    buckets: operationBuckets,
  });

  const embeddingDuration = getOrCreateHistogram(registry, {
    name: `${opts.prefix}_embedding_duration_seconds`,
    help: 'Duration of embedding function calls in seconds',
    labelNames: ['cache_name'],
    buckets: operationBuckets,
  });

  const costSavedTotal = getOrCreateCounter(registry, {
    name: `${opts.prefix}_cost_saved_total`,
    help: 'Estimated cost saved in dollars from semantic cache hits',
    labelNames: ['cache_name', 'category'],
  });

  const embeddingCacheTotal = getOrCreateCounter(registry, {
    name: `${opts.prefix}_embedding_cache_total`,
    help: 'Total embedding cache lookups (hit or miss)',
    labelNames: ['cache_name', 'result'],
  });

  const staleModelEvictions = getOrCreateCounter(registry, {
    name: `${opts.prefix}_stale_model_evictions_total`,
    help: 'Entries evicted due to staleAfterModelChange detection',
    labelNames: ['cache_name'],
  });

  const discoveryWriteFailed = getOrCreateCounter(registry, {
    name: `${opts.prefix}_discovery_write_failed_total`,
    help: 'Count of failed discovery-marker writes (best-effort HGET/HSET/SET operations against __betterdb:* keys)',
    labelNames: ['cache_name'],
  });

  const configRefreshFailed = getOrCreateCounter(registry, {
    name: `${opts.prefix}_config_refresh_failed_total`,
    help: 'Count of failed periodic config refreshes (HGETALL on __config).',
    labelNames: ['cache_name'],
  });

  const judgeDecisions = getOrCreateCounter(registry, {
    name: `${opts.prefix}_judge_decisions_total`,
    help: 'LLM-as-judge decisions for borderline cache hits',
    labelNames: ['cache_name', 'category', 'decision'],
  });

  const judgeDuration = getOrCreateHistogram(registry, {
    name: `${opts.prefix}_judge_duration_seconds`,
    help: 'Wall-clock duration of judgeFn invocations',
    labelNames: ['cache_name', 'category', 'decision'],
    buckets: [0.05, 0.1, 0.25, 0.5, 1, 2, 5],
  });

  return {
    tracer,
    metrics: {
      requestsTotal,
      similarityScore,
      operationDuration,
      embeddingDuration,
      costSavedTotal,
      embeddingCacheTotal,
      staleModelEvictions,
      discoveryWriteFailed,
      configRefreshFailed,
      judgeDecisions,
      judgeDuration,
    },
  };
}
